src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs

```diff
@@ -73,7 +73,14 @@ private async Task ProcessMeasurementGroups(IEnumerable<IMeasurementGroup> measu
                             }
                             catch (Exception ex)
                             {
-                                if (!Options.ExceptionService.HandleException(ex, log))
+                                // capture the message payload for error message service
+                                JArray payload = new JArray();
+                                foreach (var measurement in m.Data)
+                                {
+                                    payload.Add(JObject.Parse(System.Text.Encoding.Default.GetString(measurement.Payload.ToArray())));

```

Comment:
I notice that we are doing quite a bit of converting data in several places in the code. It makes me wonder if you could just provide multiple overloads for HandleException (one that takes type EventData and one that takes type IEnumerable<IMeasurementGroup>). You could just pass the event or measurements to the HandleException method and either do the conversions in the method or add multiple constructors to ErrorMessage and handle all conversions in one class. It would definitely make it easier to track down where data is being converted.

Reply:
removed the conversion here. now we just pass the event object to the error message service. we can convert it there if needed depending on what destination we choose